packages/next/src/client/components/router-reducer/reducers/navigate-reducer.ts

```diff
@@ -413,22 +413,32 @@ function navigateReducer_PPR(
           }
 
           if (
-            prefetchEntryCacheStatus !== PrefetchCacheEntryStatus.stale &&
             // This is just a paranoid check. When PPR is enabled, the server
             // will always send back a static response that's rendered from
             // the root. If for some reason it doesn't, we fall back to the
             // non-PPR implementation.
+            // TODO: We should get rid of the else branch and do all navigations
+            // via updateCacheNodeOnNavigation. The current structure is just
+            // an incremental step.
             flightDataPath.length === 3
           ) {
             const prefetchedTree: FlightRouterState = flightDataPath[0]
             const seedData = flightDataPath[1]
             const head = flightDataPath[2]
+
+            // Check whether the prefetched data is stale. If so, we'll
+            // ignore it and wait for the dynamic data to stream in before
+            // showing new segments.
+            const isPrefetchStale =
+              prefetchEntryCacheStatus === PrefetchCacheEntryStatus.stale
+
             const task = updateCacheNodeOnNavigation(
               currentCache,
               currentTree,
               prefetchedTree,
               seedData,
-              head
+              head,
+              isPrefetchStale
```

Comment:
The fix is that I moved the stale check into this branch and reimplemented the `addRefetchToLeafSegments` logic inside `updateCacheNodeOnNavigation`. The idea is that we shouldn't show the prefetch data, even temporarily, if it's too stale.

Reply:
Or at least that makes it match the current Next.js behavior, not sure it's ideal longer term.